cache.go

```diff
@@ -433,30 +482,43 @@ func (c *InMemoryCache[K, V]) Exists(key K) bool {
 	}
 
 	shard := c.getShard(key)
-	now := time.Now().UnixNano()
-
-	shard.mu.Lock()
-	defer shard.mu.Unlock()
+	shard.mu.RLock()
 
 	item, exists := shard.data[key]
 	if !exists {
+		shard.mu.RUnlock()
 		return false
 	}
 
-	if item.expireTime > 0 && now > item.expireTime {
-		delete(shard.data, key)
-		c.removeFromLRU(shard, item)
-		if c.config.EvictionPolicy == LFU && item.heapIndex != -1 {
-			heap.Remove(shard.lfuHeap, item.heapIndex)
-		}
-		c.itemPool.Put(item)
-		atomic.AddInt64(&shard.size, -1)
-		if c.config.StatsEnabled {
-			atomic.AddInt64(&shard.expirations, 1)
+	// Fast path for non-expiring items
+	if item.expireTime == 0 {
+		shard.mu.RUnlock()
+		return true
+	}
+
+	// For expiring items, check expiration with cached timestamp
+	now := atomic.LoadInt64(&c.cachedTime)
+	if now > item.expireTime {
+		shard.mu.RUnlock()
+		shard.mu.Lock()
```

Comment:
I would have inverted them to reduce the risk of having the resource getting unlocked

But I'm unsure it works